[WPE][browserperfdash-benchmark] Allow to run run-benchmark plans subtest-by-subtest

clopez opened the pull request:
#### 00cc05fc4770372af4aae513615ea623c0feb3c3
<pre>
[WPE][browserperfdash-benchmark] Allow to run run-benchmark plans subtest-by-subtest
<a href="https://bugs.webkit.org/show_bug.cgi?id=295050">https://bugs.webkit.org/show_bug.cgi?id=295050</a>

Reviewed by Nikolas Zimmermann.

On the RPi4 32-bit bots, JetStream2 frequently crashes or times out before completing,
which prevents any results from being reported because the current runner only uploads
data if all subtests finish successfully.

We have experimented with running selected subtest sets, but the flakiness remains:
some subtests pass while others fail inconsistently, which makes very difficult to
select a working set of subtests.

This patch implements on browserperfdash-benchmark the ability to run a given benchmark
plan subtest-by-subtest, so it only runs a subtest at a time.
If a subtest fails, the runner skips it and proceeds to the next. Partial results from the
passing subtests are then uploaded to the dashboard, improving visibility into progress
and regressions even when the full benchmark cannot complete.

To differenciate the standard benchmark plan between the one that is run subtest-by-subtest
the string `-split-subtests` is appended to the end of the benchmark plan name. So, to
to run the plan jetstream2 subtest-by-subtest it should be specified to run a virtual
benchmark plan named `jetstream2-split-subtests`.
Since the benchmark plan name is visible on the dashboard, is also possible to differentiate
there from the complete jetstream2 vs the one that was run subtest-by-subtest

All the benchmark plans that support subtests can be specified this way (not only JetStream2),
and the list of those is visible when the flag `--list-plans` is passed to the runner.

* Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py:
(BrowserPerfDashRunner.__init__):
(BrowserPerfDashRunner._parse_config_file):
(BrowserPerfDashRunner._get_plan_version_hash):
(BrowserPerfDashRunner._get_benchmark_runner_split_subtess_plans):
(BrowserPerfDashRunner._run_benchmark_runner_plan_split_subtests):
(BrowserPerfDashRunner._run_plan):
(BrowserPerfDashRunner.run):

Canonical link: <a href="https://commits.webkit.org/296877@main">https://commits.webkit.org/296877@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b430de0e21773e3dd34b116be4a5b915866c9d9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109866 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29524 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19954 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/115887 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/60103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111829 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30202 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38112 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/83520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/60103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112814 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/24076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/98939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/63961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/109298 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/23452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/17086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/59682 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/93446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/17129 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/118679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36905 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/27351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/118679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37278 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/95205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/118679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37296 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/15036 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/32751 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17728 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36800 "Built successfully") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/36460 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/39802 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/38169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

aoikonomopoulos commented on Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py:
```diff
@@ -342,11 +348,88 @@ def _run_benchmark_runner_plan(self, plan, result_file_path):
+        subtests_passed = []
+        subtests_failed = []
+        for subtest in subtests_to_run:
+            current_subtest_index += 1
```
Could this use `enumerate()` instead?

aoikonomopoulos commented on Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py:
```diff
@@ -342,11 +348,88 @@ def _run_benchmark_runner_plan(self, plan, result_file_path):
+                    metrics_aggregation_entry = temp_result_json[main_benchmark_name]['metrics']
+                    split_subtest_data[main_benchmark_name] = {}
+                    split_subtest_data[main_benchmark_name]['metrics'] = metrics_aggregation_entry
+                    split_subtest_data[main_benchmark_name]['tests'] = {}
```
Another style thing, but perhaps we could construct the dict locally and then expose it? I.e.
``` python
split_subtest_data[main_benchmark_name] = {
    'metrics': temp_result_json[main_benchmark_name]['metrics'],
    'tests' = {},
}
```

It's fine the way it is, it's just that the repetition of the long names makes it feel a tad too verbose for me.

aoikonomopoulos commented:
Left some (mostly stylistic) suggestions, conceptually this looks solid and is sorely needed!

clopez commented on Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py:
```diff
@@ -342,11 +348,88 @@ def _run_benchmark_runner_plan(self, plan, result_file_path):
+        main_benchmark_name = None
+        metrics_aggregation_entry = None
+        split_subtest_data = {}
+        split_subtest_data_is_empty = True
```
fixed

clopez committed:
[WPE][browserperfdash-benchmark] Allow to run run-benchmark plans subtest-by-subtest
https://bugs.webkit.org/show_bug.cgi?id=295050

Reviewed by Nikolas Zimmermann.

On the RPi4 32-bit bots, JetStream2 frequently crashes or times out before completing,
which prevents any results from being reported because the current runner only uploads
data if all subtests finish successfully.

We have experimented with running selected subtest sets, but the flakiness remains:
some subtests pass while others fail inconsistently, which makes very difficult to
select a working set of subtests.

This patch implements on browserperfdash-benchmark the ability to run a given benchmark
plan subtest-by-subtest, so it only runs a subtest at a time.
If a subtest fails, the runner skips it and proceeds to the next. Partial results from the
passing subtests are then uploaded to the dashboard, improving visibility into progress
and regressions even when the full benchmark cannot complete.

To differenciate the standard benchmark plan between the one that is run subtest-by-subtest
the string `-split-subtests` is appended to the end of the benchmark plan name. So, to
to run the plan jetstream2 subtest-by-subtest it should be specified to run a virtual
benchmark plan named `jetstream2-split-subtests`.
Since the benchmark plan name is visible on the dashboard, is also possible to differentiate
there from the complete jetstream2 vs the one that was run subtest-by-subtest

All the benchmark plans that support subtests can be specified this way (not only JetStream2),
and the list of those is visible when the flag `--list-plans` is passed to the runner.

* Tools/Scripts/webkitpy/browserperfdash/browserperfdash_runner.py:
(BrowserPerfDashRunner.__init__):
(BrowserPerfDashRunner._parse_config_file):
(BrowserPerfDashRunner._get_plan_version_hash):
(BrowserPerfDashRunner._get_benchmark_runner_split_subtess_plans):
(BrowserPerfDashRunner._run_benchmark_runner_plan_split_subtests):
(BrowserPerfDashRunner._run_plan):
(BrowserPerfDashRunner.run):

Canonical link: https://commits.webkit.org/296877@main
